truffleruby/src/main/java/org/truffleruby/cext/CExtNodes.java

```diff
@@ -1085,4 +1087,32 @@ protected ObjectIVarSetNode createObjectIVarSetNode() {
 
     }
 
+    @CoreMethod(names = "store_exception", onSingleton = true, needsBlock = true)
+    public abstract static class StoreExceptionNode extends YieldingCoreMethodNode {
+
+        @Specialization
+        public Object executeWithProtect(VirtualFrame frame, DynamicObject block) {
+            try {
+                yield(frame, block);
+                return nil();
+            } catch (Throwable e) {
+                return e;
+            }
+        }
+    }
+
+    @CoreMethod(names = "raise_exception", onSingleton = true, required = 1)
+    public abstract static class RaiseExceptionNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        public Object executeWithProtect(Throwable e) {
+            if (e instanceof Error) {
+                throw (Error) e;
+            } else if (e instanceof RuntimeException) {
+                throw (RuntimeException) e;
+            }
+            throw new RuntimeException(e);
+        }
```

Comment:
Yes, this is super annoying in Java, I had to do the same recently for Mutex:
https://github.com/graalvm/truffleruby/blob/7300169ef4b0de9ad508b2e4991db15a803af735/truffleruby/src/main/java/org/truffleruby/core/mutex/MutexOperations.java#L80-L88

The only difference is I used a JavaException to wrap checked exceptions, like in other places of the code base.
We should probably share this in something like ExceptionOperations.rethrow(Throwable).
It's also possible to throw a checked exception, but that is dark magic.